File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int height(TreeNode root){
        if(root==null){
            return -1;
        }
        int left = height(root.left);
        int right =  height(root.right);
        
        int h = Math.max(left,right)+1;
        return h;
        
    }
    public int diameterOfBinaryTree(TreeNode root) {
        if(root==null){
            return 0;
        }
        int left = diameterOfBinaryTree(root.left);
        int right =  diameterOfBinaryTree(root.right);
        int sd = height(root.left) + height(root.right)+2;
        return Math.max(left,Math.max(right,sd));
        
        
        
        
        
    }
}

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    pair<int,int> diameterFast(TreeNode* root){
        if(root == NULL){
            pair<int,int> ans = make_pair(0,0);
            return ans;
        }
        pair<int,int> left = diameterFast(root->left);
        pair<int,int> right = diameterFast(root->right);
        int op1 = left.first;
        int op2 = right.first;
        int op3 = left.second + right.second + 1;
        pair<int , int> res ;
        res.first = max(op1,max(op2,op3));
        res.second = max(left.second , right.second) +1;
        return res;
    }
    int diameterOfBinaryTree(TreeNode* root) {
        return diameterFast(root).first-1;
    }
};